harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -308,8 +313,14 @@ def _train_for_step(
                     metric = metric.cpu().detach().numpy()
                 tr_metrics[name] = metric
 
-            per_batch_metrics.append(tr_metrics)
+            epoch_per_batch_metrics.append(tr_metrics)
+            if self.context.is_epoch_end():
+                for callback in self.callbacks.values():
+                    callback.on_training_epoch_end(epoch_per_batch_metrics)
+                per_batch_metrics += epoch_per_batch_metrics
+                epoch_per_batch_metrics = []
 
+        per_batch_metrics += epoch_per_batch_metrics
```

Comment:
This isn't right. If a step has batches in two sequential epochs, `per_batch_metrics` contains the metrics of all the batches but `epoch_per_batch_metrics` would clear itself once it hits an epoch end.

Reply:
👍🏼  we removed support for on_train_epoch_end since we can't guarantee epochwide metrics for this callback with mid-epoch preemption